src/data/real/liouville.lean

```diff
@@ -28,34 +24,143 @@ A Liouville number is a real number `x` such that for every natural number `n`,
 begin
+  -- By contradiction, `x = a / b`, with `a ∈ ℤ`, `0 < b ∈ ℕ` is a Liouville number,
   rintros ⟨⟨a, b, bN0, cop⟩, rfl⟩,
-  change (is_liouville (a / b)) at h,
+  -- clear up the mess of constructions of rationals
+  change (liouville (a / b)) at h,
+  -- Since `a / b` is a Liouville number, there are `p, q ∈ ℤ`, with `q1 : 1 < q`,
+  -- `a0 : a / b ≠ p / q` and `a1 : abs (a / b - p / q) < 1 / q ^ (b + 1)`
   rcases h (b + 1) with ⟨p, q, q1, a0, a1⟩,
+  -- A few useful inequalities
   have qR0 : (0 : ℝ) < q := int.cast_pos.mpr (zero_lt_one.trans q1),
   have b0 : (b : ℝ) ≠ 0 := ne_of_gt (nat.cast_pos.mpr bN0),
   have bq0 : (0 : ℝ) < b * q := mul_pos (nat.cast_pos.mpr bN0) qR0,
+  -- At a1, clear denominators...
   replace a1 : abs (a * q - b * p) * q ^ (b + 1) < b * q, by
     rwa [div_sub_div _ _ b0 (ne_of_gt qR0), abs_div, div_lt_div_iff (abs_pos.mpr (ne_of_gt bq0))
-      (pow_pos qR0 _), abs_of_pos bq0, one_mul, ← int.cast_pow, ← int.cast_mul, ← int.cast_coe_nat,
-      ← int.cast_mul, ← int.cast_mul, ← int.cast_sub, ← int.cast_abs, ← int.cast_mul, int.cast_lt]
-        at a1,
+      (pow_pos qR0 _), abs_of_pos bq0, one_mul,
+      -- ... and revert to integers
+       ← int.cast_pow, ← int.cast_mul, ← int.cast_coe_nat, ← int.cast_mul, ← int.cast_mul,
+       ← int.cast_sub, ← int.cast_abs, ← int.cast_mul, int.cast_lt] at a1,
+  -- At a0, clear denominators...
   replace a0 : ¬a * q - ↑b * p = 0, by
     rwa [ne.def, div_eq_div_iff b0 (ne_of_gt qR0), mul_comm ↑p, ← sub_eq_zero_iff_eq,
+      -- ... and revert to integers
       ← int.cast_coe_nat, ← int.cast_mul, ← int.cast_mul, ← int.cast_sub, int.cast_eq_zero] at a0,
+  -- Actually, `q` is a natural number
   lift q to ℕ using (zero_lt_one.trans q1).le,
+  -- Looks innocuous, but we now have an integer with non-zero absolute value: this is at
+  --least one away from zero.  The gain here is what gets the proof going.
   have ap : 0 < abs (a * ↑q - ↑b * p) := abs_pos.mpr a0,
+  -- Actually, the absolute value of an integer is a natural number
   lift (abs (a * ↑q - ↑b * p)) to ℕ using (abs_nonneg (a * ↑q - ↑b * p)),
+  -- At a1, revert to natural numbers
   rw [← int.coe_nat_mul, ← int.coe_nat_pow, ← int.coe_nat_mul, int.coe_nat_lt] at a1,
+  -- Recall this is by contradiction: we obtained the inequality `b * q ≤ x * q ^ (b + 1)`, so
+  -- we are done.
   exact not_le.mpr a1 (nat.mul_lt_mul_pow_succ (int.coe_nat_pos.mp ap) (int.coe_nat_lt.mp q1)).le,
 end
 
-end is_liouville
+end liouville
+
+open polynomial metric set real ring_hom
+
+/-- This lemma collects the properties needed to prove `exists_pos_real_of_irrational_root`.
+It is stated in more general form than needed: in the intended application, `Z = ℤ`, `N = ℕ`,
+`R = ℝ`, `d a = a ^ f.nat_degree`, `j z a  = z / (a + 1)`, `f ∈ ℤ[x]`, `α` is an irrational
+root of `f`, `ε` is small, `M` is a bound on the Lipschitz constant of `f` near `α`, `n` is
+the degree of the polynomial `f`.
```

Comment:
When reading the docstring, I have no clue about the content of the lemma. The docstring should explain the content of the lemma (this is missing) and explain the context and motivation (it is there, and very nicely).

Reply:
I have tried (in my head) explaining the content of this lemma, but I find it hard to formulate a "wordy" version that is not directly a transcription of the statement.  The main reason for formulating this lemma is that it helps formalizing the next one.  It avoids coercions kicking in, simplifies the manipulation of terms in the algebraic expressions and replaces certain steps with hypotheses which are direct consequences of lemmas in mathlib.

Here is a very verbose attempt at an explanation: what do you think?

```
/-- Let `Z, N` be types, let `R` be a metric space, let `α : R` be a point and let
`j : Z → N → R` be a function.  We aim to estimate how close we can get to `α`, while staying
in the image of `j`.  The points `j z a` of `R` in the image of `j` come with a "cost" equal to
`d a`.  As we get closer to `α` while staying in the image of `j`, we are interested in bounding
the quantity `d a * dist α (j z a)` from below by a strictly positive amount `1 / M`: the intuition
is that approximating `α` with the points in the image of `j` should come at a high cost.  The
hypotheses on the function `f : R → R` provide us with sufficient conditions to ensure our goal.
The first hypothesis is that `f` is Lipschitz at `α`: this yields a bound on the distance.
The second hypothesis is specific to the Liouville argument and provides the missing bound
involving the cost function `d`.

This lemma collects the properties needed to prove `exists_pos_real_of_irrational_root`.
It is stated in more general form than needed: in the intended application, `Z = ℤ`, `N = ℕ`,
`R = ℝ`, `d a = a ^ f.nat_degree`, `j z a  = z / (a + 1)`, `f ∈ ℤ[x]`, `α` is an irrational
root of `f`, `ε` is small, `M` is a bound on the Lipschitz constant of `f` near `α`, `n` is
the degree of the polynomial `f`.
-/
```